<a href="https://colab.research.google.com/github/Hasan-Iqtedar/Netflix-Challenge/blob/master/Netflix_Challenge_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import gc
import scipy
from scipy.sparse import coo_matrix
from sklearn.model_selection import train_test_split
from sklearn.decomposition import NMF




In [ ]:
import pandas as pd
import math

In [2]:
df1 = pd.read_csv("/content/drive/My Drive/combined_data_1.txt", header = None, names = ['Cust_Id', 'Rating'], usecols = [0,1] )

In [3]:
df1['Rating'] = df1['Rating'].astype(float)


In [4]:
print('Dataset 1 shape: {}'.format(df1.shape))

print(df1[0:5])

Dataset 1 shape: (24058263, 2)
   Cust_Id  Rating
0       1:     NaN
1  1488844     3.0
2   822109     5.0
3   885013     4.0
4    30878     4.0


In [4]:
df2 = pd.read_csv("/content/drive/My Drive/combined_data_2.txt", header = None, names = ['Cust_Id', 'Rating'], usecols = [0,1] )
df3 = pd.read_csv("/content/drive/My Drive/combined_data_3.txt", header = None, names = ['Cust_Id', 'Rating'], usecols = [0,1] )
df4 = pd.read_csv("/content/drive/My Drive/combined_data_4.txt", header = None, names = ['Cust_Id', 'Rating'], usecols = [0,1] )

df2['Rating'] = df2['Rating'].astype(float)
df3['Rating'] = df3['Rating'].astype(float)
df4['Rating'] = df4['Rating'].astype(float)

print('Dataset 2 shape: {}'.format(df2.shape))
print('Dataset 3 shape: {}'.format(df3.shape))
print('Dataset 4 shape: {}'.format(df4.shape))

#print('\n',df2[0:5],'\n\n')
#print('\n',df3[0:5],'\n\n')
#print('\n',df4[0:5],'\n\n')



Dataset 2 shape: (26982302, 2)
Dataset 3 shape: (22605786, 2)
Dataset 4 shape: (26851926, 2)


In [5]:

df = df1.append(df2, ignore_index = True)
df = df.append(df3, ignore_index = True)
df = df.append(df4, ignore_index = True)

del df1,df2,df3,df4

In [8]:
#gc.collect()
movie_count = df.isnull().sum()[1]
print('Total movies: ',movie_count)
print("Dataset shape: {}".format(df.shape))
del m1,m2,m3,m4,m

Total movies:  17770
Dataset shape: (100498277, 2)


In [9]:
print(df.iloc[::5000000, :])
#print(df1.iloc[50:55])

           Cust_Id  Rating
0               1:     NaN
5000000    2560324     4.0
10000000   2271935     2.0
15000000   1921803     2.0
20000000   1933327     3.0
25000000   1465002     3.0
30000000    961023     4.0
35000000   1372532     5.0
40000000    854274     5.0
45000000    116334     3.0
50000000    768483     3.0
55000000   1331144     5.0
60000000   1609324     2.0
65000000   1699240     3.0
70000000   1776418     4.0
75000000   1643826     5.0
80000000    932047     4.0
85000000   2292868     4.0
90000000    932191     4.0
95000000   1815101     3.0
100000000   872339     4.0


In [7]:
#Removing rows containing movie id
df = df[pd.notnull(df['Rating'])]
print(df.iloc[::5000000, :])

           Cust_Id  Rating
1          1488844     3.0
5000996     501954     2.0
10001962    404654     5.0
15002876    886608     2.0
20003825   1193835     2.0
25004661   1899206     3.0
30005496    154804     4.0
35006274   2078749     5.0
40007057    450763     5.0
45007991    102092     3.0
50009023    220298     5.0
55010042    550530     5.0
60011038    222570     3.0
65011875   1273080     5.0
70012676   2026970     5.0
75013582    506044     4.0
80014453    353605     2.0
85015116    664606     3.0
90016008   2213715     3.0
95016879   1589401     5.0
100017627  2314006     4.0


In [9]:
gc.collect()

0

In [11]:
print(df[0:5])

   Cust_Id  Rating
1  1488844     3.0
2   822109     5.0
3   885013     4.0
4    30878     4.0
5   823519     3.0


In [12]:
print(df.shape)

(100480507, 2)


In [6]:
#Adding movie ids to corresponding rows in the dataframe so as to have a cleaned dataset.
#This is achieved by iterating through all the files and making a list of movie ids for
#the respective customer id and rating line.
#Then the whole list is added as a column to the main dataframe to get the desired dataset.


dataSet = ['/content/drive/My Drive/combined_data_1.txt',
           '/content/drive/My Drive/combined_data_2.txt',
           '/content/drive/My Drive/combined_data_3.txt',
           '/content/drive/My Drive/combined_data_4.txt',
          ]

x = 1
movie_id = 0
movie = []

for file in dataSet:

  handle = open(file)
#Start of inner for loop.
  for line in handle:
    if line.startswith(str(x)+":"):
      movie_id = x
      x += 1
      continue
    movie.append(movie_id)
#End of inner for loop.  
  handle.close()
  gc.collect()
#End of outer for loop.

print(df.shape,'\n',len(movie))


(100498277, 2) 
 100480507


In [10]:
df['Movies'] = movie
del movie

In [11]:
gc.collect()

28

In [18]:
print(df.iloc[::5000000, :])
#print(df1.iloc[50:55])

           Cust_Id  Rating  Movies
1          1488844     3.0       1
5000996     501954     2.0     996
10001962    404654     5.0    1962
15002876    886608     2.0    2876
20003825   1193835     2.0    3825
25004661   1899206     3.0    4661
30005496    154804     4.0    5496
35006274   2078749     5.0    6274
40007057    450763     5.0    7057
45007991    102092     3.0    7991
50009023    220298     5.0    9023
55010042    550530     5.0   10042
60011038    222570     3.0   11038
65011875   1273080     5.0   11875
70012676   2026970     5.0   12676
75013582    506044     4.0   13582
80014453    353605     2.0   14453
85015116    664606     3.0   15116
90016008   2213715     3.0   16008
95016879   1589401     5.0   16879
100017627  2314006     4.0   17627


In [13]:
#Installing surprise package to use the SVD function.
!pip install surprise

     |████████████████████████████████| 11.8MB 260kB/s 
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.1-cp36-cp36m-linux_x86_64.whl size=1670943 sha256=0db1b2797fb1d0d4ed2a38ae3aefcaeffcb83ab3b2cff2b4377f2f79da0ce4c5
  Stored in directory: /root/.cache/pip/wheels/78/9c/3d/41b419c9d2aff5b6e2b4c0fc8d25c538202834058f9ed110d0
Successfully built scikit-surprise


In [14]:
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate

In [15]:
gc.collect()

3373

In [ ]:

reader = Reader()

# get just top 100K rows for faster run time
data = Dataset.load_from_df(df[['Cust_Id', 'Movies', 'Rating']], reader)
#data.split(n_folds=3)

svd = SVD(n_epochs=5, verbose = True)
cross_validate(svd, data, measures=['RMSE', 'MAE'])


The SVD function in the surprise library has an improved cost function implementation and hence would result in better rmse. But due to the enormity of the dataset, it cannot run on colab and running the above cell crashes the runtime. 


> 


So, to train a model for collaborative filtering, matrix factorization is used below using the h2o4gpu library. But the cost function in this library is traditional one and will not result in optimum rmse.

## **Collaborative Filtering using Matrix Factorization**

Restart runtime before running and import the libraries in the first cell.
**Testing scipy sparse matrix**

In [ ]:
#By specifying a particular row and column and then the corresponding value
#in another array, we can construct a sparse matrix using scipy.
#The rows and columns which are not defined in the arrays have values zero.

rows = [0,1,2,3,1,0]#Specifying rows.
cols = [0,1,2,3,3,3]#Specifying columns.
vals = [3,4,20,19,99,1]#Specifying corresponding values e.g value at row 1 colum 3 is 99.

vals = np.array(vals, np.float32)
cols = np.array(cols, np.int32)
rows = np.array(rows, np.int32)

#Constructing the sparse matrix. 
R = coo_matrix( ( vals, (rows,cols) ), shape = (4,4) )#.toarray()
print(R.toarray())
#type(R)
#print('\n',R[0][:]*R[1][:].T)
print("nice")


[[ 3.  0.  0.  1.]
 [ 0.  4.  0. 99.]
 [ 0.  0. 20.  0.]
 [ 0.  0.  0. 19.]]
nice


**np**.**unique**

In [ ]:
#np.unique returns the unique elements of the array (sorted). If return_inverse
#is true then it also returns the index form of the given array in which each
#element is replaced by its index in the unique array.
x = [1,2,3,4,4,3,2]
x = np.array(x, np.int32)
ar, indices = np.unique(x, return_inverse = True)

print('unique sorted x: ', ar)
print('index form of original x: ', indices)



unique sorted x:  [1 2 3 4]
index form of original x:  [0 1 2 3 3 2 1]


**Using Scikit Learn for Matrix Factorization**

In [ ]:
#from sklearn.decomposition import NMF
'''
R = [[ 3.  0.  0.  1.]
 [ 0.  4.  0. 99.]
 [ 0.  0. 20.  0.]
 [ 0.  0.  0. 19.]]
R = np.array(R)
'''
print(R,"\n\n")
nmf = NMF(1000, init='random', max_iter=2000, random_state=0,tol=0.001)

W = nmf.fit_transform(R);
H = nmf.components_;
nR = np.dot(W,H)
print (nR)

  (0, 0)	3.0
  (1, 1)	4.0
  (2, 2)	20.0
  (3, 3)	19.0
  (1, 3)	99.0
  (0, 3)	1.0 


[[3.00000000e+00 1.12460815e-16 0.00000000e+00 1.00000000e+00]
 [0.00000000e+00 4.00000000e+00 0.00000000e+00 9.90000000e+01]
 [0.00000000e+00 4.14832442e-16 2.00000000e+01 2.90436701e-15]
 [8.78143080e-16 4.60196649e-01 0.00000000e+00 1.90000000e+01]]


In [ ]:
print(W.shape,'\n',H.shape)

(4, 1000) 
 (1000, 4)


Scikit learn does not deal with the missing values (zeros) and predicts them to be zeros. The purpose of matrix factorization for collaborative filtering is to get these missing values which are the unknown ratings. So, we cannot use scikit learn.

In [ ]:
#Installing h2o4gpu.
!apt-get install -y libopenblas-dev pbzip2
!pip install -U tabulate==0.8.2
!pip install h2o4gpu
import h2o4gpu

Reading package lists... Done
Building dependency tree       
Reading state information... Done
libopenblas-dev is already the newest version (0.2.20+ds-4).
pbzip2 is already the newest version (1.1.9-1build1).
The following package was automatically installed and is no longer required:
  libnvidia-common-440
Use 'apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 39 not upgraded.
Requirement already up-to-date: tabulate==0.8.2 in /usr/local/lib/python3.6/dist-packages (0.8.2)


**Using h2o4gpu for Matrix Factorization**

In [ ]:
#from sklearn.model_selection import train_test_split

#from scipy.sparse import csr_matrix
'''
R = [
     [5,3,0,1],
     [4,0,0,1],
     [1,1,0,5],
     [1,0,0,4],
     [0,1,5,4],
    ]
#R = np.array(R)
#print(R)
#R = coo_matrix(R)
'''

rows = [0,0,0, 1,1, 2,2,2, 3,3, 4,4,4]
cols = [0,1,3, 0,3, 0,1,3, 0,3, 1,2,3]
vals = [5,3,1, 4,1, 1,1,5, 1,4, 1,5,4]

vals = np.array(vals, dtype= np.float32)
cols = np.array(cols, dtype= np.int32)
rows = np.array(rows)

R = coo_matrix((vals ,(rows, cols)), shape = (5,4)  )

#####Make sparse matrix from scratch#####

print("Splitting the data\n")
print(type(R))

train_row, test_row, train_col, test_col, train_data, test_data = train_test_split(R.row, R.col, R.data, test_size = 0.2, random_state = 44)
train = coo_matrix( (train_data, (train_row, train_col)), shape = (5,4) )
test = coo_matrix( (test_data, (test_row, test_col)), shape = (5,4) )

print("R:\n",R.toarray())
print("\nR_train:\n", train.toarray())
print("\nR_test:\n", test.toarray())
type(train)
#type(test)

#del train,test,R

Splitting the data

<class 'scipy.sparse.coo.coo_matrix'>
R:
 [[5. 3. 0. 1.]
 [4. 0. 0. 1.]
 [1. 1. 0. 5.]
 [1. 0. 0. 4.]
 [0. 1. 5. 4.]]

R_train:
 [[5. 3. 0. 1.]
 [4. 0. 0. 1.]
 [0. 1. 0. 0.]
 [1. 0. 0. 4.]
 [0. 0. 5. 4.]]

R_test:
 [[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [1. 0. 0. 5.]
 [0. 0. 0. 0.]
 [0. 1. 0. 0.]]


scipy.sparse.coo.coo_matrix

In [ ]:

features = 30
batches = 1
lmbda = 0.1
scores = []

factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter = 20)

print(factorization.dtype)

factorization.fit( train, X_test= test, verbose = True, early_stopping_rounds= 5 )


type(factorization.thetaT)
print(factorization.thetaT.shape,'\n',factorization.XT.shape)


<class 'numpy.float32'>
iteration 0 train: 0.36212393641471863 cv: 2.5662426948547363
iteration 1 train: 0.24573847651481628 cv: 2.4999542236328125
iteration 2 train: 0.2048846036195755 cv: 2.679933786392212
iteration 3 train: 0.1821260303258896 cv: 2.696864128112793
iteration 4 train: 0.1680992990732193 cv: 2.7147746086120605
iteration 5 train: 0.159223735332489 cv: 2.756368398666382
iteration 6 train: 0.15344418585300446 cv: 2.7689859867095947
iteration 7 train: 0.14871618151664734 cv: 2.7762818336486816
best iteration:1 train: 0.24573847651481628 cv: 2.4999542236328125
(4, 30) 
 (5, 30)


In [ ]:
print('Approximated R:\n\n', np.dot(factorization.XT, factorization.thetaT.T) )


Approximated R:

 [[4.7711444  2.867989   1.6241543  0.9808193 ]
 [3.8261504  2.2702065  1.5287249  0.9716469 ]
 [1.5567601  0.949642   0.42473698 0.23387425]
 [0.9429118  0.00581822 4.682519   3.751868  ]
 [2.0354273  0.6880902  4.860801   3.817715  ]]


So, h2o4gpu does deal with the missing values and so is suitable for collaborative filtering.

In [ ]:
#import gc
#import numpy as np
#import scipy
#from sklearn.model_selection import train_test_split

dataset = [
    '/content/drive/My Drive/combined_data_1.txt',
    '/content/drive/My Drive/combined_data_2.txt',
    '/content/drive/My Drive/combined_data_3.txt',
    '/content/drive/My Drive/combined_data_4.txt',
]

rows = []
columns = []
values = []

for file in dataset:
  
    print('processing ',file)
    with open(file, "r") as data:
        movie = -1
        for line in data:
            if line.endswith(':\n'):
                movie = int(line.split(':')[0])-1   #int(line[:-2]) - 1
                continue
            assert movie >= 0
            line_as_list = line.split(',')
            user = int(line_as_list[0])
            rating = float(line_as_list[1])
            rows.append(user)
            columns.append(movie)
            values.append(rating)
    gc.collect()

print('Forming scipy.sparse.coo_matrix')

values = np.array(values, dtype=np.float32)
columns = np.array(columns, dtype=np.int32)
rows = np.array(rows)

#np.unique returns the unique elements of the array (sorted). If return_inverse
#is true then it also returns the index form of the given array in which each
#element is replaced by its index in the unique array.

user, indices = np.unique(rows, return_inverse=True)
user = user.astype(np.int32)

gc.collect()

#Constructing the sparse matrix.
user_item_matrix = scipy.sparse.coo_matrix((values, (indices, columns)))

shape = user_item_matrix.shape
print('User-Item matrix size', shape)

gc.collect()

processing  /content/drive/My Drive/combined_data_1.txt
processing  /content/drive/My Drive/combined_data_2.txt
processing  /content/drive/My Drive/combined_data_3.txt
processing  /content/drive/My Drive/combined_data_4.txt
Forming scipy.sparse.coo_matrix
User-Item matrix size (480189, 17770)


0

In [ ]:
print(type(user_item_matrix))

<class 'scipy.sparse.coo.coo_matrix'>


In [ ]:
print('splitting into training and test set')
train_row, test_row, train_col, test_col, train_data, test_data = train_test_split(
    user_item_matrix.row, user_item_matrix.col, user_item_matrix.data, test_size=0.2, random_state=44)

train = scipy.sparse.coo_matrix( (train_data, (train_row, train_col)), shape=shape)
test = scipy.sparse.coo_matrix( (test_data, (test_row, test_col)), shape=shape)

splitting into training and test set


In [ ]:
print("Train shape: ",train.shape,"\nTest shape: ",test.shape)

Train shape:  (480189, 17770) 
Test shape:  (480189, 17770)


In [ ]:
type(test)

scipy.sparse.coo.coo_matrix

In [ ]:
features = 10

lmbda = 1 #Regularization parameter lambda.
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES = 1



scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 1.3794498443603516 cv: 1.3828246593475342
iteration 1 train: 1.362408995628357 cv: 1.3655656576156616
iteration 2 train: 1.3596627712249756 cv: 1.3628290891647339
iteration 3 train: 1.3591103553771973 cv: 1.3622788190841675
iteration 4 train: 1.3589965105056763 cv: 1.3621654510498047
iteration 5 train: 1.358972191810608 cv: 1.3621413707733154
iteration 6 train: 1.3589680194854736 cv: 1.3621371984481812
iteration 7 train: 1.358967661857605 cv: 1.362136721611023
iteration 8 train: 1.3589680194854736 cv: 1.3621370792388916
iteration 9 train: 1.3589677810668945 cv: 1.362136721611023
iteration 10 train: 1.3589677810668945 cv: 1.3621368408203125
iteration 11 train: 1.3589677810668945 cv: 1.3621368408203125
iteration 12 train: 1.3589677810668945 cv: 1.3621368408203125
iteration 13 train: 1.358967900276184 cv: 1.362136960029602
best iteration:7 train: 1.358967661857605 cv: 1.362136721611023
best iteration: 7


In [ ]:
features = 10

lmbda = 0.9
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1



scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 1.2966309785842896 cv: 1.300516128540039
iteration 1 train: 1.288570523262024 cv: 1.2921522855758667
iteration 2 train: 1.287410020828247 cv: 1.2909936904907227
iteration 3 train: 1.2871254682540894 cv: 1.290710210800171
iteration 4 train: 1.287052035331726 cv: 1.2906370162963867
iteration 5 train: 1.2870326042175293 cv: 1.290617823600769
iteration 6 train: 1.2870279550552368 cv: 1.290613055229187
iteration 7 train: 1.2870270013809204 cv: 1.2906122207641602
iteration 8 train: 1.287027359008789 cv: 1.2906125783920288
iteration 9 train: 1.2870272397994995 cv: 1.2906124591827393
iteration 10 train: 1.2870272397994995 cv: 1.2906124591827393
iteration 11 train: 1.2870272397994995 cv: 1.2906124591827393
iteration 12 train: 1.2870272397994995 cv: 1.2906124591827393
iteration 13 train: 1.2870272397994995 cv: 1.2906124591827393
best iteration:7 train: 1.2870270013809204 cv: 1.2906122207641602
best iteration: 7


**Features = 10, Lambda = 0.1**


In [ ]:
features = 10

lmbda = 0.1
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1



scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 0.9209076762199402 cv: 0.9342684149742126
iteration 1 train: 0.8983924388885498 cv: 0.9096971750259399
iteration 2 train: 0.8691772222518921 cv: 0.8835955858230591
iteration 3 train: 0.8589157462120056 cv: 0.8751086592674255
iteration 4 train: 0.8531450629234314 cv: 0.870635986328125
iteration 5 train: 0.849221408367157 cv: 0.867718517780304
iteration 6 train: 0.8467674255371094 cv: 0.8659597635269165
iteration 7 train: 0.845411479473114 cv: 0.8650335669517517
iteration 8 train: 0.8446244597434998 cv: 0.8645270466804504
iteration 9 train: 0.8440980911254883 cv: 0.864207923412323
iteration 10 train: 0.8437188267707825 cv: 0.8639912605285645
iteration 11 train: 0.8434510827064514 cv: 0.8638489246368408
iteration 12 train: 0.8432750105857849 cv: 0.863765299320221
iteration 13 train: 0.8431692719459534 cv: 0.8637251257896423
iteration 14 train: 0.8431124091148376 cv: 0.863713800907135
iteration 15 train: 0.8430876135826111 cv: 0.863720715045929
iteration 16 train: 0.8430

**Features = 20, Lambda = 0.1**

In [ ]:
features = 20

lmbda = 0.1
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1



scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 0.922046959400177 cv: 0.941326379776001
iteration 1 train: 0.8939289450645447 cv: 0.9083305597305298
iteration 2 train: 0.8679646849632263 cv: 0.8847139477729797
iteration 3 train: 0.858080267906189 cv: 0.8763487339019775
iteration 4 train: 0.852331817150116 cv: 0.8717412948608398
iteration 5 train: 0.8487703800201416 cv: 0.8689836859703064
iteration 6 train: 0.8467338681221008 cv: 0.8674196004867554
iteration 7 train: 0.8454866409301758 cv: 0.8664594292640686
iteration 8 train: 0.8446215391159058 cv: 0.8657926917076111
iteration 9 train: 0.8439647555351257 cv: 0.8652862906455994
iteration 10 train: 0.84344083070755 cv: 0.8648819923400879
iteration 11 train: 0.8430122137069702 cv: 0.8645504117012024
iteration 12 train: 0.8426559567451477 cv: 0.864273726940155
iteration 13 train: 0.8423569798469543 cv: 0.8640406727790833
iteration 14 train: 0.8421044945716858 cv: 0.8638429641723633
iteration 15 train: 0.8418903946876526 cv: 0.8636742830276489
iteration 16 train: 0.841

**Features = 30, Lambda = 0.1**

In [ ]:
features = 30

lmbda = 0.1
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1

scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 0.9271710515022278 cv: 0.9509572386741638
iteration 1 train: 0.8925923705101013 cv: 0.9093681573867798
iteration 2 train: 0.8699245452880859 cv: 0.8883312940597534
iteration 3 train: 0.8594346642494202 cv: 0.8792544603347778
iteration 4 train: 0.853863000869751 cv: 0.8745435476303101
iteration 5 train: 0.8504490256309509 cv: 0.8716713786125183
iteration 6 train: 0.8482115864753723 cv: 0.8697755932807922
iteration 7 train: 0.8466642498970032 cv: 0.8684459328651428
iteration 8 train: 0.8455336689949036 cv: 0.8674584031105042
iteration 9 train: 0.8446744084358215 cv: 0.8666942715644836
iteration 10 train: 0.8440025448799133 cv: 0.8660857677459717
iteration 11 train: 0.8434651494026184 cv: 0.8655906319618225
iteration 12 train: 0.8430265188217163 cv: 0.8651809692382812
iteration 13 train: 0.8426626324653625 cv: 0.8648372292518616
iteration 14 train: 0.8423563838005066 cv: 0.8645455837249756
iteration 15 train: 0.8420956134796143 cv: 0.8642961978912354
iteration 16 train:

**Features = 10, Lambda = 0.09**

In [ ]:
features = 10

lmbda = 0.09
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1

scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 0.919082522392273 cv: 0.9337408542633057
iteration 1 train: 0.8912507891654968 cv: 0.9045393466949463
iteration 2 train: 0.86146479845047 cv: 0.8783158659934998
iteration 3 train: 0.8503574728965759 cv: 0.8692879676818848
iteration 4 train: 0.8436897993087769 cv: 0.8641850352287292
iteration 5 train: 0.8396382927894592 cv: 0.8612208366394043
iteration 6 train: 0.8375617861747742 cv: 0.8597744703292847
iteration 7 train: 0.8364052772521973 cv: 0.8590083122253418
iteration 8 train: 0.8356488347053528 cv: 0.8585264086723328
iteration 9 train: 0.8351474404335022 cv: 0.8582193851470947
iteration 10 train: 0.8348405957221985 cv: 0.8580418825149536
iteration 11 train: 0.8346686959266663 cv: 0.857952356338501
iteration 12 train: 0.8345800638198853 cv: 0.8579162955284119
iteration 13 train: 0.8345388770103455 cv: 0.8579100966453552
iteration 14 train: 0.8345237374305725 cv: 0.8579197525978088
iteration 15 train: 0.8345234990119934 cv: 0.8579378128051758
iteration 16 train: 0.

**Features = 20, Lambda = 0.09**

In [ ]:
features = 20

lmbda = 0.09
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1

scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 0.9176719188690186 cv: 0.9394670128822327
iteration 1 train: 0.8838629722595215 cv: 0.9014418721199036
iteration 2 train: 0.8572823405265808 cv: 0.8776885271072388
iteration 3 train: 0.8464813232421875 cv: 0.868797242641449
iteration 4 train: 0.84029221534729 cv: 0.8640092015266418
iteration 5 train: 0.8369607925415039 cv: 0.8615228533744812
iteration 6 train: 0.8350040912628174 cv: 0.8600958585739136
iteration 7 train: 0.8336581587791443 cv: 0.8591333627700806
iteration 8 train: 0.8326548337936401 cv: 0.8584248423576355
iteration 9 train: 0.831879734992981 cv: 0.8578796982765198
iteration 10 train: 0.831265389919281 cv: 0.8574479222297668
iteration 11 train: 0.8307672739028931 cv: 0.8570973873138428
iteration 12 train: 0.8303550481796265 cv: 0.8568066358566284
iteration 13 train: 0.8300079703330994 cv: 0.8565616607666016
iteration 14 train: 0.8297115564346313 cv: 0.8563523888587952
iteration 15 train: 0.8294562101364136 cv: 0.8561720848083496
iteration 16 train: 0.8

**Features = 40, Lambda = 0.09**

In [ ]:
features = 40

lmbda = 0.09
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1

scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 0.9290470480918884 cv: 0.9604634046554565
iteration 1 train: 0.8859197497367859 cv: 0.9085109233856201
iteration 2 train: 0.8599783182144165 cv: 0.8842960596084595
iteration 3 train: 0.8490181565284729 cv: 0.8745591640472412
iteration 4 train: 0.8429808020591736 cv: 0.8693076968193054
iteration 5 train: 0.8392906785011292 cv: 0.8660891652107239
iteration 6 train: 0.836770236492157 cv: 0.8638776540756226
iteration 7 train: 0.8349432945251465 cv: 0.8622585535049438
iteration 8 train: 0.8335690498352051 cv: 0.8610249161720276
iteration 9 train: 0.8325038552284241 cv: 0.8600568175315857
iteration 10 train: 0.8316574096679688 cv: 0.8592787384986877
iteration 11 train: 0.8309704661369324 cv: 0.8586417436599731
iteration 12 train: 0.8304030299186707 cv: 0.8581116795539856
iteration 13 train: 0.8299270868301392 cv: 0.8576653003692627
iteration 14 train: 0.8295233845710754 cv: 0.85728520154953
iteration 15 train: 0.8291776776313782 cv: 0.8569592237472534
iteration 16 train: 0

**Features = 50, Lambda = 0.09**

In [ ]:
features = 50

lmbda = 0.09
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1

scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 0.9375385642051697 cv: 0.972289502620697
iteration 1 train: 0.887265145778656 cv: 0.9119730591773987
iteration 2 train: 0.8620395660400391 cv: 0.8878561854362488
iteration 3 train: 0.8508060574531555 cv: 0.8774559497833252
iteration 4 train: 0.844757616519928 cv: 0.8718321323394775
iteration 5 train: 0.8408340215682983 cv: 0.8681830167770386
iteration 6 train: 0.8380693197250366 cv: 0.8656010031700134
iteration 7 train: 0.8360319137573242 cv: 0.8636833429336548
iteration 8 train: 0.8344753980636597 cv: 0.8622075915336609
iteration 9 train: 0.8332513570785522 cv: 0.8610408306121826
iteration 10 train: 0.832267701625824 cv: 0.860099196434021
iteration 11 train: 0.8314639329910278 cv: 0.8593271970748901
iteration 12 train: 0.8307993412017822 cv: 0.8586863279342651
iteration 13 train: 0.8302444219589233 cv: 0.8581491708755493
iteration 14 train: 0.8297770023345947 cv: 0.857694685459137
iteration 15 train: 0.829380452632904 cv: 0.8573076725006104
iteration 16 train: 0.829

**Features = 10, Lambda = 0.01**

In [ ]:
features = 10

lmbda = 0.01
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=1

scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 0.9023697376251221 cv: 0.9483940601348877
iteration 1 train: 0.847931981086731 cv: 0.9051058888435364
iteration 2 train: 0.8170322179794312 cv: 0.8747328519821167
iteration 3 train: 0.8035961985588074 cv: 0.8613352179527283
iteration 4 train: 0.7973046898841858 cv: 0.8551712036132812
iteration 5 train: 0.7941975593566895 cv: 0.8521674275398254
iteration 6 train: 0.7925224900245667 cv: 0.850562334060669
iteration 7 train: 0.791532576084137 cv: 0.8496137261390686
iteration 8 train: 0.7909086346626282 cv: 0.8490219116210938
iteration 9 train: 0.7904977798461914 cv: 0.8486401438713074
iteration 10 train: 0.7902175784111023 cv: 0.8483867049217224
iteration 11 train: 0.7900206446647644 cv: 0.8482202887535095
iteration 12 train: 0.7898778915405273 cv: 0.8481069207191467
iteration 13 train: 0.7897714376449585 cv: 0.8480294942855835
iteration 14 train: 0.7896901965141296 cv: 0.8479746580123901
iteration 15 train: 0.7896267175674438 cv: 0.8479333519935608
iteration 16 train: 0

In [ ]:
features = 100

lmbda = 0.09
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=15

scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 1.01876962184906 cv: 1.0621757507324219
iteration 1 train: 0.9216195940971375 cv: 0.9510647058486938
iteration 2 train: 0.8839120268821716 cv: 0.9126104116439819
iteration 3 train: 0.8656641840934753 cv: 0.8942202925682068
iteration 4 train: 0.854899525642395 cv: 0.8834402561187744
iteration 5 train: 0.8479428291320801 cv: 0.8764787912368774
iteration 6 train: 0.8432057499885559 cv: 0.8717188239097595
iteration 7 train: 0.8398457169532776 cv: 0.8683204054832458
iteration 8 train: 0.8373790383338928 cv: 0.8658079504966736
iteration 9 train: 0.8355146646499634 cv: 0.8638963103294373
iteration 10 train: 0.8340698480606079 cv: 0.8624061346054077
iteration 11 train: 0.8329264521598816 cv: 0.8612208962440491
iteration 12 train: 0.8320049047470093 cv: 0.860261857509613
iteration 13 train: 0.8312508463859558 cv: 0.8594747185707092
iteration 14 train: 0.8306258320808411 cv: 0.8588203191757202
iteration 15 train: 0.8301019072532654 cv: 0.8582707643508911
iteration 16 train: 0.

In [ ]:
features = 150

lmbda = 0.1
# increase it in case out-of GPU memory, but n_components / BATCHES has to be a multiple of 10
BATCHES=20

scores = []
factorization = h2o4gpu.solvers.FactorizationH2O(features, lmbda, max_iter=50)
factorization.fit(train, X_test=test, X_BATCHES=BATCHES, THETA_BATCHES=BATCHES, scores=scores, verbose=True, early_stopping_rounds=5)
print('best iteration:',factorization.best_iteration)

iteration 0 train: 1.1704909801483154 cv: 1.2095153331756592
iteration 1 train: 0.9815908670425415 cv: 1.006775140762329
iteration 2 train: 0.9198583960533142 cv: 0.9438983201980591
iteration 3 train: 0.8917080163955688 cv: 0.9153193235397339
iteration 4 train: 0.8759885430335999 cv: 0.8993964195251465
iteration 5 train: 0.8663532137870789 cv: 0.8896118402481079
iteration 6 train: 0.8600542545318604 cv: 0.8831852674484253
iteration 7 train: 0.8557230234146118 cv: 0.8787420392036438
iteration 8 train: 0.8526203632354736 cv: 0.8755417466163635
iteration 9 train: 0.8503212332725525 cv: 0.8731583952903748
iteration 10 train: 0.8485693335533142 cv: 0.8713340163230896
iteration 11 train: 0.8472029566764832 cv: 0.8699056506156921
iteration 12 train: 0.846116304397583 cv: 0.8687660098075867
iteration 13 train: 0.8452378511428833 cv: 0.8678422570228577
iteration 14 train: 0.8445178270339966 cv: 0.8670834898948669
iteration 15 train: 0.8439208269119263 cv: 0.8664531707763672
iteration 16 train: 

factorization.XT is the users matrix: users x features

factorization.thetaT is the movies matrix: movies x features

Multiplying the two matrices gives the reconstructed user-item matrix which contains the ratings for non-rated movies.



In [ ]:
print(factorization.XT.shape,'\n',factorization.XT.dtype)
print(factorization.thetaT.shape)

(480189, 150) 
 float32
(17770, 150)


From above observations, we get good results when **features = 10** and **lambda = 0.09** 

best iteration: 13          

**train error (mse):** 0.8345388770103455  

**validation error (mse):** 0.8579100966453552

Which quite fairly low variance and low bias.

In [ ]:
type(factorization.XT)

numpy.ndarray